### PR TITLE
repair & improve names generated from structs' definitions

### LINKED ISSUE #930: struct-copy fails accross modules
### What version of Racket are you using?
7.7.0.2
### What program did you run?
```
#lang typed/racket/base

(module mod1 typed/racket/base
  (provide (all-defined-out))
  (struct a ([aa : Number]) #:type-name A)
  (define a1 (a 1))
  (struct-copy a a1 [aa 2]))

(module mod2 typed/racket/base
  (require (submod ".." mod1))
  (define a2 (a 1))
  (struct-copy a a2 [aa 2]))
```
### What should have happened?
create new struct a

### If you got an error message, please include it here.
struct-copy: accessor name not associated with the given structure type in: a-aa

probably related to #322

### PR DESCRIPTION
**Pull request by capfredf**
Previously when a struct's type name was different from the struct name, TR
did not attach its syntax transformer binding with the corresponding struct
info for the enclosing module, but did so for its exported transformer
binding. This was the root cause of the problems mentioned of #930.

With this change, the behavior is consistent, the type name of a structure can
be used as an alternative to its struct id. E.g. we can use a structure type name
in struct-copy, or use it as the super id in other structs' defintions.

Under the hood, this change includes a fair amount of refactoring:

1. Move existing helper functions and structures for struct-info from the
"jack-of-all-trades" utils/utils. Add more functions and structures. In particular, we
nows use three struct-info structures to handle three different situations of
generating bindings of a struct's type name and struct name.

2. Where the binding of a structure's type name for the enclosing module is
generated has moved in favor of the consistency.

3. The heavy-lifting job of provide handling has moved to def-binding.rkt, where
every binding structure takes care of its corresponding quad generation.

lastly, closes #930

### CODE REVIEW

File: typed-racket-lib/typed-racket/utils/utils.rkt [hunk 1]
```diff
@@ -232,16 +232,21 @@ at least theoretically.
...
-            (type-name-error stx)))
-      #:property prop:struct-info (λ (x) (extract-struct-info (struct-info-self-ctor-info x))))
-    (struct-info-self-ctor id info flag)))
+(struct struct-info-self-ctor (id info type)
+  #:property prop:procedure
+  (lambda (ins stx)
+    (self-ctor-transformer (struct-info-self-ctor-id ins) stx))
+  #:property prop:struct-info (λ (x) (extract-struct-info (struct-info-self-ctor-info x))))
+
+(define (get-type-from-struct-info ins)
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/utils.rkt:241]: Need a signature/purpose.

File: typed-racket-lib/typed-racket/utils/utils.rkt [hunk 2]
```diff
@@ -232,16 +232,21 @@ at least theoretically.
...
-  (let ()
-    (struct struct-info-self-ctor (id info type-is-constr?)
-      #:property prop:procedure
-      (lambda (ins stx)
-        (if (struct-info-self-ctor-type-is-constr? ins)
-            (self-ctor-transformer (struct-info-self-ctor-id ins) stx)
-            (type-name-error stx)))
-      #:property prop:struct-info (λ (x) (extract-struct-info (struct-info-self-ctor-info x))))
-    (struct-info-self-ctor id info flag)))
+(struct struct-info-self-ctor (id info type)
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/utils.rkt:235]: Need a data definition.

File: typed-racket-test/succeed/gh-issue-930.rkt [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#lang typed/racket/base
```
Comment (samth) [inline typed-racket-test/succeed/gh-issue-930.rkt:1]: Need tests for all the things that worked previously (ie, using a type name a struct name, which was a bug).

File: typed-racket-lib/typed-racket/utils/struct-info-helper.rkt [hunk 4]
```diff
@@ -0,0 +1,148 @@
+#lang racket/base
```
Comment (samth) [inline typed-racket-lib/typed-racket/utils/struct-info-helper.rkt:1]: I think `-helper` is not needed in the file name.

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 5]
```diff
@@ -117,19 +118,30 @@
...
-                                  (tc-error/stx #'v.par "parent type not a valid structure name: ~a"
-                                                (syntax->datum #'v.par)))))])
-                (values #'v.name parent0 parent))
-          (values #'v.name #f #f))]))
+     (if (attribute v.par)
+         (let* ([parent-si-ins (syntax-local-value #'v.par (lambda () #f))]
+                [parent0 (parse-type
+                          (cond
+                            ;; maybe-struct-info-wrapper-type only returns parent's
+                            ;; type name when parent-si-ins is an TR's
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:126]: I'm not sure exactly what this comment means. Is it that this case happens when the structure is defined in a `#lang typed/racket` module?

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 6]
```diff
@@ -117,19 +118,30 @@
...
+                            ;; struct-info-self-ctor instance.
+                            [(maybe-struct-info-wrapper-type parent-si-ins)]
+                            ;; When parent-si-ins is either simply #f or
+                            ;; a struct-info-self-ctor instance from untyped racket,
+                            ;; i.e. the parent is a builtin structure like exn,
+                            ;; the parent structure name is also the type name.
+                            [else #'v.par]))]
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:133]: Is this only for builtin structs, or also for `require/typed`?

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt (hunk 5 above)
Comment (samth) [reply typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:126]: I think this also happens for `require/typed`.

File: typed-racket-lib/typed-racket/static-contracts/combinators/struct.rkt [hunk 7]
```diff
@@ -37,14 +37,16 @@
...
+          #`(struct/c #,name #,@(map f args))]))
      (define (sc->constraints v f)
        (match v
         [(struct-combinator args _ mut?)
-         (merge-restricts*
-           (if mut? 'chaperone 'flat)
-           (map (lambda (a) (if (not mut?) (add-constraint (f a) 'chaperone) (f a))) args))]))])
+         (merge-restricts* (if mut? 'chaperone 'flat)
+                           (map (lambda (a)
+                                  (if (not mut?) (add-constraint (f a) 'chaperone)
```
Comment (samth) [inline typed-racket-lib/typed-racket/static-contracts/combinators/struct.rkt:47]: `if` should be 3 lines or 1.

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 8]
```diff
@@ -117,19 +118,30 @@
...
-                                 (else
-                                  (tc-error/stx #'v.par "parent type not a valid structure name: ~a"
-                                                (syntax->datum #'v.par)))))])
-                (values #'v.name parent0 parent))
-          (values #'v.name #f #f))]))
+     (if (attribute v.par)
+         (let* ([parent-si-ins (syntax-local-value #'v.par (lambda () #f))]
+                [parent0 (parse-type
+                          (cond
+                            ;; maybe-struct-info-wrapper-type only returns parent's
```
Comment (capfredf) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:125]: also for require/type

File: typed-racket-lib/typed-racket/typecheck/provide-handling.rkt [hunk 9]
```diff
@@ -44,161 +37,20 @@
...
-    ([(internal-id external-ids) (in-sorted-free-id-table provs)])
-    (define-values (defs export-def id alias) (mk internal-id))
-    (define provide-forms
-      (for/list ([external-id (in-list external-ids)])
-        #`(rename-out [#,id #,external-id])))
-    (values #`(begin #,defs)
-            export-def
-            #`(provide #,@provide-forms)
-            alias)))
+  (start-making-quads def-tbl pos-blame-id mk-redirect-id
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/provide-handling.rkt:45]:
```suggestion
  (parameterize ([... def-tbl] [... pos-blame-id] [... mk-redirect-id])
    (with-fresh-mapping
```

File: typed-racket-lib/typed-racket/typecheck/def-binding.rkt [hunk 10]
```diff
@@ -1,19 +1,198 @@
...
+  ((make-syntax-introducer) id))
+
+(define-generics providable
+  [mk-quad providable new-id])
+
+(define (mk-ignored-quad i)
+  (values #'(begin) #'(begin) i null))
+
+;; the parameters for a module being checked.
+(define mapping (make-parameter #f))
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/def-binding.rkt:31]: Add a comment explaining what `mapping` is.

File: typed-racket-lib/typed-racket/typecheck/provide-handling.rkt [hunk 11]
```diff
@@ -44,161 +37,20 @@
...
+                      (for/lists (defs^ export-defs provides aliases)
+                                 ;; sort provs to generate deterministic output
+                                 ([(internal-id external-ids) (in-sorted-free-id-table provs)])
+                        (define-values (defs^ export-def id alias) (make-quad internal-id))
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/provide-handling.rkt:49]:
```suggestion
                        (define-values (defs^ export-def id alias) (make-quad internal-id def-tbl pos-blame-id mk-redirect-id))
```

Comment (samth) [reply typed-racket-lib/typed-racket/typecheck/provide-handling.rkt:49]: And then the parameterize goes away.

**Review by samth — APPROVED** [review]

**Pull request merged**